File: frontend/lib/api/folders.ts
/**
 * Folders API client for FastAPI backend.
 * Uses HttpOnly cookies for authentication.
 */

import type { Folder } from "../types"

const API_BASE = "/api/backend/folders"

export interface ApiError {
  detail: string
}

export interface FolderCreateResponse {
  success: boolean
  count: number
}

export interface FolderUpdateResponse {
  success: boolean
  message?: string
}

export interface FolderDeleteResponse {
  success: boolean
}

/**
 * Transform backend snake_case folder to frontend camelCase.
 */
function transformFolder(raw: Record<string, unknown>): Folder {
  return {
    id: raw.id as string,
    name: raw.name as string,
    order: (raw.order as number) ?? 0,
    createdAt: raw.created_at ? new Date(raw.created_at as string) : new Date(),
  }
}

/**
 * Transform frontend camelCase folder to backend snake_case.
 */
function toApiFormat(folder: Partial<Folder>): Record<string, unknown> {
  const result: Record<string, unknown> = {}

  if (folder.id !== undefined) result.id = folder.id
  if (folder.name !== undefined) result.name = folder.name
  if (folder.order !== undefined) result.order = folder.order
  if (folder.createdAt !== undefined) {
    result.created_at = folder.createdAt instanceof Date
      ? folder.createdAt.toISOString()
      : folder.createdAt
  }

  return result
}

/**
 * Get all folders for the authenticated user.
 */
export async function getFolders(): Promise<Folder[]> {
  const response = await fetch(API_BASE, {
    method: "GET",
    credentials: "include",
  })

  if (!response.ok) {
    const error: ApiError = await response.json()
    throw new Error(error.detail || "Failed to get folders")
  }

  const data = await response.json()
  return data.map(transformFolder)
}

/**
 * Create or upsert multiple folders.
 */
export async function saveFolders(folders: Partial<Folder>[]): Promise<FolderCreateResponse> {
  const response = await fetch(API_BASE, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify(folders.map(toApiFormat)),
  })

  if (!response.ok) {
    const error: ApiError = await response.json()
    if (response.status === 409) {
      throw new Error("duplicate")
    }
    throw new Error(error.detail || "Failed to save folders")
  }

  return response.json()
}

/**
 * Update a folder by ID.
 * Supports partial updates - only provided fields will be updated.
 */
export async function updateFolder(
  folderId: string,
  updates: Partial<Folder>
): Promise<FolderUpdateResponse> {
  const response = await fetch(`${API_BASE}/${folderId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify(toApiFormat(updates)),
  })

  if (!response.ok) {
    const error: ApiError = await response.json()
    throw new Error(error.detail || "Failed to update folder")
  }

  return response.json()
}

/**
 * Delete a folder.
 * Note: Feeds in this folder will have their folder_id set to null.
 */
export async function deleteFolder(folderId: string): Promise<FolderDeleteResponse> {
  const response = await fetch(`${API_BASE}/${folderId}`, {
    method: "DELETE",
    credentials: "include",
  })

  if (!response.ok) {
    const error: ApiError = await response.json()
    throw new Error(error.detail || "Failed to delete folder")
  }

  return response.json()
}

/**
 * Folders API namespace for easy import.
 */
export const foldersApi = {
  getFolders,
  saveFolders,
  updateFolder,
  deleteFolder,
}

File: backend/app/api/routers/folders.py
"""Folders API router for CRUD operations."""

import logging
from typing import List
from uuid import UUID
from fastapi import APIRouter, Depends

from app.dependencies import create_service_dependency, require_exists, extract_update_data
from app.exceptions import DuplicateError, ValidationError
from app.schemas.folders import (
    FolderCreate,
    FolderCreateWithId,
    FolderUpdate,
    FolderResponse,
)
from app.services.db.folders import FolderService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/folders", tags=["folders"])


get_folder_service = create_service_dependency(FolderService)


@router.get("", response_model=List[FolderResponse])
async def get_folders(service: FolderService = Depends(get_folder_service)):
    """
    Get all folders for the authenticated user.

    Returns:
        List of folders ordered by order field.
    """
    folders = service.load_folders()
    logger.debug(f"Retrieved {len(folders)} folders")
    return folders


@router.post("", response_model=dict)
async def create_folders(
    folders: List[FolderCreateWithId],
    service: FolderService = Depends(get_folder_service),
):
    """
    Create or upsert multiple folders.

    Supports bulk creation/update of folders.

    Args:
        folders: List of folders to create/update

    Returns:
        Success status with count.
    """
    folder_dicts = [folder.model_dump() for folder in folders]
    result = service.save_folders(folder_dicts)

    if not result.get("success"):
        error = result.get("error", "Unknown error")
        if error == "duplicate":
            raise DuplicateError("folder name")
        raise ValidationError(error)

    logger.info(f"Created/updated {len(folders)} folders")
    return {"success": True, "count": len(folders)}


@router.put("/{folder_id}", response_model=dict)
async def update_folder(
    folder_id: UUID,
    folder_update: FolderUpdate,
    service: FolderService = Depends(get_folder_service),
):
    """
    Update a folder by ID.

    Supports partial updates - only provided fields will be updated.

    Args:
        folder_id: UUID of the folder to update
        folder_update: Fields to update

    Returns:
        Success status.

    Raises:
        404 if folder not found.
    """
    require_exists(service.get_folder(str(folder_id)), "Folder")

    update_data = extract_update_data(folder_update)

    if not update_data:
        return {"success": True, "message": "No fields to update"}

    result = service.update_folder(str(folder_id), update_data)

    if not result.get("success"):
        error = result.get("error", "Unknown error")
        raise ValidationError(error)

    logger.info(f"Updated folder {folder_id}")
    return {"success": True}


@router.delete("/{folder_id}", response_model=dict)
async def delete_folder(
    folder_id: UUID,
    service: FolderService = Depends(get_folder_service),
):
    """
    Delete a folder.

    Note: Feeds in this folder wave their folder_id set to null.

    Args:
        folder_id: UUID of the folder to delete

    Returns:
        Success status.

    Raises:
        404 if folder not found.
    """
    require_exists(service.get_folder(str(folder_id)), "Folder")

    service.delete_folder(str(folder_id))
    logger.info(f"Deleted folder {folder_id}")
    return {"success": True}
